src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -61,63 +60,91 @@ public List<GitCommit> GetCommitHistory(string file, string committish = null)
             }
         }
 
-        private List<GitCommit> GetCommitHistory(GitCommitCache.FileCommitCache commitCache, string file, string committish = null)
+        public void Save()
         {
-            const int MaxParentBlob = 32;
+            if (_commitCache.IsValueCreated)
+            {
+                _commitCache.Value.Save();
+            }
+        }
 
-            var (commits, commitsBySha) = _commits.GetOrAdd(
-                committish ?? "",
-                key => new Lazy<(List<Commit>, Dictionary<long, Commit>)>(() => LoadCommits(key))).Value;
+        public void Dispose()
+        {
+            var repo = Interlocked.Exchange(ref _repo, IntPtr.Zero);
+            if (repo != IntPtr.Zero)
+            {
+                git_repository_free(_repo);
+                GC.SuppressFinalize(this);
+            }
+        }
 
-            if (commits.Count <= 0)
+        ~FileCommitProvider()
+        {
+            Dispose();
+        }
+
+        private unsafe GitCommit[] GetCommitHistory(GitCommitCache.FileCommitCache commitCache, string file, string committish = null)
+        {
+            if (git_revparse_single(out var pHead, _repo, committish ?? "HEAD") != 0)
             {
-                return new List<GitCommit>();
+                throw Errors.CommittishNotFound(_repoPath, committish).ToException();
             }
 
+            var (commits, parents, closeNodes, openNodes) = _objectPool.TryTake(out var pooled)
```

Comment:
if we just want to reduce the allocation, can we initialize `commits`, `closeNodes`, `openNodes` as private field?